// ngx-tetris/esm2020/lib/tetris-core.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TetrisCoreComponent } from './tetris-core.component';
import { BoardComponent } from './board/board.component';
import { GameManagerService } from './services/game-manager.service';
import { TileComponent } from './tile/tile.component';
import * as i0 from "@angular/core";
export class TetrisCoreModule {
}
TetrisCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: TetrisCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TetrisCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: TetrisCoreModule, declarations: [TetrisCoreComponent,
        BoardComponent,
        TileComponent], imports: [CommonModule], exports: [TetrisCoreComponent] });
TetrisCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: TetrisCoreModule, providers: [
        GameManagerService
    ], imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: TetrisCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        TetrisCoreComponent,
                        BoardComponent,
                        TileComponent
                    ],
                    providers: [
                        GameManagerService
                    ],
                    exports: [TetrisCoreComponent]
                }]
        }] });
